1110-i386-txt-Add-Intel-TXT-core-implementation.patch

```diff
@@ -0,0 +1,1038 @@
++      heap_end_element->size = sizeof (*heap_end_element);
++    }
++  grub_dprintf ("slaunch", "TXT HEAP init done\n");
++  /*
++   * TODO: TXT spec: Note: BiosDataSize + OsMleDataSize + OsSinitDataSize + SinitMleDataSize
++   * must be less than or equal to TXT.HEAP.SIZE, TXT spec, p. 102.
++   */
++
++  return GRUB_ERR_NONE;
++}
++
++/*
++ * TODO: Why 1 GiB limit? It does not seem that it is required by TXT spec.
++ * If there is a limit then it should be checked before allocation and image load.
++ *
++ * If enough room is available in front of the MLE, the maximum size of an
++ * MLE that can be covered is 1G. This is due to having 512 PDEs pointing
++ * to 512 page tables with 512 PTEs each.
++ */
++grub_uint32_t
++grub_txt_get_mle_ptab_size (grub_uint32_t mle_size)
++{
++  /*
++   * #PT + 1 PT + #PD + 1 PD + 1 PDT
++   *
++   * Why do we need 2 extra PTEs and PDEs? Yes, because MLE image may not
++   * start and end at PTE (page) and PDE (2 MiB) boundary...
++   */
++  return ((((mle_size / GRUB_PAGE_SIZE) + 2) / 512) + 1 +
++	  (((mle_size / (512 * GRUB_PAGE_SIZE)) + 2) / 512) + 1 + 1) * GRUB_PAGE_SIZE;
++}
++
++/* Page directory and table entries only need Present set */
++#define MAKE_PT_MLE_ENTRY(addr)  (((grub_uint64_t)(grub_addr_t)(addr) & GRUB_PAGE_MASK) | 0x01)
++
++/*
++ * The MLE page tables have to be below the MLE and have no special regions in
++ * between them and the MLE (this is a bit of an unwritten rule).
++ * 20 pages are carved out of memory below the MLE. That leave 18 page table
++ * pages that can cover up to 36M .
++ * can only contain 4k pages
++ *
++ * TODO: TXT Spec p.32; List section name and number with PT MLE requirments here.
++ *
++ * TODO: This function is not able to cover MLEs larger than 1 GiB. Fix it!!!
++ * After fixing inrease GRUB_TXT_MLE_MAX_SIZE too.
++ */
++void
++grub_txt_setup_mle_ptab (struct grub_slaunch_params *slparams)
++{
++  grub_uint8_t *pg_dir, *pg_dir_ptr_tab = slparams->mle_ptab_mem, *pg_tab;
++  grub_uint32_t mle_off = 0, pd_off = 0;
++  grub_uint64_t *pde, *pte;
++
++  grub_memset (pg_dir_ptr_tab, 0, slparams->mle_ptab_size);
++
++  pg_dir         = pg_dir_ptr_tab + GRUB_PAGE_SIZE;
++  pg_tab         = pg_dir + GRUB_PAGE_SIZE;
++
++  /* Only use first entry in page dir ptr table */
++  *(grub_uint64_t *)pg_dir_ptr_tab = MAKE_PT_MLE_ENTRY(pg_dir);
++
++  /* Start with first entry in page dir */
++  *(grub_uint64_t *)pg_dir = MAKE_PT_MLE_ENTRY(pg_tab);
++
++  pte = (grub_uint64_t *)pg_tab;
++  pde = (grub_uint64_t *)pg_dir;
++
++  do
++    {
++      *pte = MAKE_PT_MLE_ENTRY(slparams->mle_start + mle_off);
++
++      pte++;
++      mle_off += GRUB_PAGE_SIZE;
++
++      if (!(++pd_off % 512))
++        {
++          /* Break if we don't need any additional page entries */
++          if (mle_off >= slparams->mle_size)
++            break;
++          pde++;
++          *pde = MAKE_PT_MLE_ENTRY(pte);
++        }
++    } while (mle_off < slparams->mle_size);
++}
```

Comment:
@HW42 @marmarek would you mind checking the page table construction code?  I assume you know more about page tables than I do.